python/lsst/ap/association/testApdb.py

```diff
@@ -0,0 +1,575 @@
+        -------
+        mergedDiaSourceHistory : `pandas.DataFrame`
+            The combined catalog, with all of the rows from preloadedDiaSources
+            catalog ordered before the rows of associatedDiaSources catalog.
+        mergedDiaObjects : `pandas.DataFrame`
+            Table of new DiaObjects merged with their history.
+        updatedDiaObjectIds : `numpy.Array`
+            Object Id's from associated diaSources.
+        """
+        # Index the DiaSource catalog for this visit after all associations
+        # have been made.
+        updatedDiaObjectIds = associatedDiaSources["diaObjectId"][
+            associatedDiaSources["diaObjectId"] != 0].to_numpy()
+        associatedDiaSources.set_index(["diaObjectId",
+                                        "band",
+                                        "diaSourceId"],
+                                       drop=False,
+                                       inplace=True)
+
+        # Append new DiaObjects and DiaSources to their previous history.
+        if diaObjects.empty:
+            mergedDiaObjects = newDiaObjects.set_index("diaObjectId", drop=False)
+        elif not newDiaObjects.empty:
+            mergedDiaObjects = pd.concat(
+                [diaObjects,
+                 newDiaObjects.set_index("diaObjectId", drop=False)],
+                sort=True)
+        else:
+            mergedDiaObjects = diaObjects
+
+        mergedDiaSourceHistory = DiaPipelineTask.mergeCatalogs(self, preloadedDiaSources,
+                                                               associatedDiaSources,
+                                                               "preloadedDiaSources")
+
+        # Finally, update the diaObject table with the number of associated diaSources
+        mergedUpdatedDiaObjects = DiaPipelineTask.updateObjectTable(mergedDiaObjects, mergedDiaSourceHistory)
+        return (mergedDiaSourceHistory, mergedUpdatedDiaObjects, updatedDiaObjectIds)
+
+    def runForcedMeasurement(self, diaObjects, updatedDiaObjects, idGenerator):
+        """Forced Source Measurement
+
+        Forced photometry on the difference and calibrated exposures using the
+        new and updated DiaObject locations.
+
+        Parameters
+        ----------
+        diaObjects : `pandas.DataFrame`
+            Catalog of DiaObjects.
+        updatedDiaObjects : `pandas.DataFrame`
+            Catalog of updated DiaObjects.
+        idGenerator : `lsst.meas.base.IdGenerator`
+            Object that generates source IDs and random number generator seeds.
+
+        Returns
+        -------
+        diaForcedSources : `pandas.DataFrame`
+            Catalog of calibrated forced photometered fluxes on both the
+            difference and direct images at DiaObject locations.
+        """
+        # Restrict forced source measurement to objects with sufficient history to be reliable.
+        objectTable = diaObjects.query(f'nDiaSources >= {self.config.historyThreshold}')
```

Comment:
Hmm, this depends on `nDiaSources` being correct and that depends on having DiaSource history.

Reply:
Yes, this is the one place I currently use the diaSource history. I will attempt to re-write it without needing that history, so that I can skip reading diaSources.